webapp/channels/src/components/color_input.tsx

```diff
@@ -13,172 +13,155 @@ type Props = {
     isDisabled?: boolean;
 }
 
-type State = {
-    focused: boolean;
-    isOpened: boolean;
-    value: string;
-}
-
-export default class ColorInput extends React.PureComponent<Props, State> {
-    private colorPicker: React.RefObject<HTMLDivElement>;
-    private colorInput: React.RefObject<HTMLInputElement>;
-
-    public constructor(props: Props) {
-        super(props);
-        this.colorPicker = React.createRef();
-        this.colorInput = React.createRef();
-
-        this.state = {
-            focused: false,
-            isOpened: false,
-            value: props.value,
-        };
-    }
-
-    static getDerivedStateFromProps(props: Props, state: State) {
-        if (!state.focused && props.value !== state.value) {
-            return {
-                value: props.value,
-            };
-        }
-
-        return null;
+const ColorInput = ({
+    id,
+    onChange: onChangeFromProps,
+    value: valueFromProps,
+    isDisabled,
+}: Props) => {
+    const colorPicker = useRef<HTMLDivElement>(null);
+    const colorInput = useRef<HTMLInputElement>(null);
+
+    const [isFocused, setIsFocused] = useState(false);
+    const [isOpened, setIsOpened] = useState(false);
+    const [valueFromState, setValueFromState] = useState(valueFromProps);
+
+    if (!isFocused && valueFromProps !== valueFromState) {
+        setValueFromState(valueFromProps);
     }
 
-    public componentDidUpdate(prevProps: Props, prevState: State) {
-        const {isOpened: prevIsOpened} = prevState;
-        const {isOpened} = this.state;
+    useEffect(() => {
+        const checkClick = (e: MouseEvent): void => {
+            if (!colorPicker.current || !colorPicker.current.contains(e.target as Element)) {
+                setIsOpened(false);
+            }
+        };
 
-        if (isOpened !== prevIsOpened) {
+        /**
+         * Since 'isOpened' is changed by a click event, this 'useEffect' runs before the screen is painted
+         * therefore, 'checkClick' is fired before the component is mounted which in turn calls 'setIsOpened(false)'.
+         * It's not possible to update the state of an unmounted component. 'setTimeout' ensures the event listener is
+         * added after the screen is painted and the component mounted. The delay (1) is just the smallest number possible
+         * so the listeners are attached as soon as possible.
+         * https://react.dev/reference/react/useEffect#:~:text=If%20your%20Effect%20is%20caused%20by%20an%20interaction%20(like%20a%20click)%2C%20React%20may%20run%20your%20Effect%20before%20the%20browser%20paints%20the%20updated%20screen.
+         * */
+        setTimeout(() => {
```

Comment:
Could we perhaps get rid of the `setTimeout` this to use a mousedown instead of a click? There's similar logic in the Global Header's `useClickOutsideRef` hook which uses that event, and it seems to work in a more straightforward way.

Alternatively, if you want to go the extra mile here, we could replace all of this if we moved the colour picker to use Floating UI since it handles the click outside for us as well as other stuff like the focus trapping logic. Take a look at the `useEmojiPicker` hook (`webapp/channels/src/components/emoji_picker/use_emoji_picker.tsx`) if you're interested in doing that.

Reply:
We've migrated everything that used React Bootstrap, Popper.js, and React Popper to Floating UI a while back, but we haven't looked for places like this that implement floating elements themselves

Reply:
Sorry, I forgot to answer that earlier. I'm fine with leaving the `setTimeout` here. Even if it looks a bit weird, it seems like it should work correctly